### PR TITLE
#10474: Migrate Where and refactor ternary

### PR DESCRIPTION
**Pull request by KalaivaniMCW**
### Ticket
Link to Github Issue #10474 

### Problem description
To implement ttnn.where

### What's changed
- move ttnn.where from ternary.hpp to eltwise/ternary
- move ternary bindings to ternary_pybind
- implement where op in C++
- removed torch api and used golden function in 5 test files

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/10055762377
- [x] Post commit models https://github.com/tenstorrent/tt-metal/actions/runs/10020626107
- [ ] Nightly fast dispatch - as main -  https://github.com/tenstorrent/tt-metal/actions/runs/10020624298
- [x] T3k frequent https://github.com/tenstorrent/tt-metal/actions/runs/10020627603
- [x] T3k demo https://github.com/tenstorrent/tt-metal/actions/runs/10020631401
- [x] T3k unit test https://github.com/tenstorrent/tt-metal/actions/runs/10020630282
- [x] Model Perf https://github.com/tenstorrent/tt-metal/actions/runs/10020628781
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "where_op.hpp"
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp:5]: spacing

File: ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        ttnn::add(queue_id, t2, t1, std::nullopt, operation::DEFAULT_OUTPUT_MEMORY_CONFIG, output_tensor);
+    } else {
+        output_tensor = ttnn::add(queue_id, t2, t1, std::nullopt, output_mem_config);
+    }
+    return output_tensor.value();
+}
+
+}
+}
+}
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp:85]: comment namespace

File: ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    uint8_t queue_id, const Tensor& predicate, const float value_true, const float value_false, const std::optional<MemoryConfig>& output_mem_config, std::optional<Tensor> output_tensor) {
+    Tensor t2 = ttnn::multiply(queue_id, ttnn::gtz(queue_id, predicate, output_mem_config), value_true, std::nullopt, output_mem_config);
+    Tensor t1 = ttnn::multiply(queue_id, ttnn::lez(queue_id, predicate, output_mem_config), value_false, std::nullopt, output_mem_config);
+    if(output_tensor.has_value()){
+        ttnn::add(queue_id, t2, t1, std::nullopt, operation::DEFAULT_OUTPUT_MEMORY_CONFIG, output_tensor);
+    } else {
+        output_tensor = ttnn::add(queue_id, t2, t1, std::nullopt, output_mem_config);
+    }
+    return output_tensor.value();
+}
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp:81]: Is there a way to avoid duplication like this?

Comment (KalaivaniMCW) [reply ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp:81]: @TT-BrianLiu yes, I think we now can. 
previously there were two api's mul( tensor * tensor) and mul_unary( tensor * scalar) so these variations had to be implemented separately. 
Recently we combined both functionalities available under one api for ttnn.multiply, we can go for a single implementation code. I will update these.

File: ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp (hunk 1 above)
Comment (KalaivaniMCW) [reply ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp:6]: addressed

File: ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp (hunk 2 above)
Comment (KalaivaniMCW) [reply ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp]: addressed

File: ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp (hunk 3 above)
Comment (KalaivaniMCW) [reply ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp]: I have added the changes

File: ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    const Tensor& predicate,
+    const Tensor& value_true,
+    const Tensor& value_false,
+    const std::optional<MemoryConfig>& output_mem_config,
+    std::optional<Tensor> output_tensor) {
+
+    return _where_op(queue_id, predicate, value_true, value_false, output_mem_config, output_tensor);
+}
+
+Tensor WhereOp::_where_v1(
```
Comment (eyonland) [inline ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp:59]: Please remove _v1.  It is unnecessary.

File: ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    const Tensor& predicate,
+    const float value_true,
+    const Tensor& value_false,
+    const std::optional<MemoryConfig>& output_mem_config,
+    std::optional<Tensor> output_tensor) {
+
+    return _where_op(queue_id, predicate, value_true, value_false, output_mem_config, output_tensor);
+}
+
+Tensor WhereOp::_where_v2(
```
Comment (eyonland) [inline ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp:70]: Please remove _v2.  It is unnecessary.

File: ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    const Tensor& predicate,
+    const Tensor& value_true,
+    const float value_false,
+    const std::optional<MemoryConfig>& output_mem_config,
+    std::optional<Tensor> output_tensor) {
+
+    return _where_op(queue_id, predicate, value_true, value_false, output_mem_config, output_tensor);
+}
+
+Tensor WhereOp::_where_v3(
```
Comment (eyonland) [inline ttnn/cpp/ttnn/operations/eltwise/ternary/where_op.cpp:81]: Please remove the _v3 it is unnecessary.

File: tests/ttnn/unit_tests/operations/test_ternary_composite.py [hunk 7]
```diff
@@ -43,8 +44,32 @@ def test_ternary_addcdiv_ttnn(input_shapes, value, device):
...
+@pytest.mark.parametrize(
+    "input_shapes",
+    (
+        (torch.Size([1, 1, 32, 32])),
+        (torch.Size([1, 1, 320, 384])),
+        (torch.Size([1, 3, 320, 384])),
+    ),
+)
+@pytest.mark.parametrize("predicate", [0, 1])
+def test_ternary_where_v1(input_shapes, predicate, device):
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/operations/test_ternary_composite.py:64]: what is v1?

**Review by eyonland — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]

**Pull request merged**